### PR TITLE
Added changes to migrate Andele to AndroidX.

### PR DESCRIPTION
**Pull request by rspl-kashyap**
**Description**
- Migrated to AndroidX

**Tests Performed**
- Manual testing on CKDS with updated lib

**Reviewer Focus**
- Everything

**Conditions for Merge**
- Approval from Larry.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -18,7 +18,7 @@ plugins {
 apply plugin: 'com.android.library'
 
 //  Versions are x.y.z and version code is xxyyzz.  Leading 0s are stripped.
-version = "0.2.1"
+version = "0.2.2"
 def versionNum = 201
```
Comment (hiq-larryschiefer) [inline build.gradle]: Need to also change `versionNum`.

File: src/main/java/com/hiqes/andele/Andele.java [hunk 2]
```diff
@@ -246,8 +246,8 @@ public static void checkAndExecute(Fragment fragment, ProtectedAction action) {
      *                         the ActionCallback to execute when granted, etc.
      */
     @SuppressWarnings({"WeakerAccess", "unused"})
-    public static void checkAndExecute(android.support.v4.app.Fragment fragment, ProtectedAction action) {
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/Andele.java:249]: My concern with doing this is now the library will not work with codebases which still utilize `android.support.v4`.  It would be better to define new APIs for the `androidx` variants while still leaving the older support lib based APIs in place.  That would effectively allow the library to support both older codebases and new ones.

File: src/main/java/com/hiqes/andele/RequestOwnerAppCompatActivity.java [hunk 3]
```diff
@@ -21,7 +21,7 @@
 import android.content.Context;
 import android.content.pm.PackageManager;
 import android.os.Build;
-import android.support.v7.app.AppCompatActivity;
+import androidx.appcompat.app.AppCompatActivity;
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/RequestOwnerAppCompatActivity.java:24]: Same comment as above regarding the fragments.  We should support both `android.support` and `androix.appcompat`.

**Review by hiq-larryschiefer — CHANGES_REQUESTED** [review]

File: src/main/java/com/hiqes/andele/Andele.java (hunk 2 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/Andele.java:249]: Both are supported now.

File: src/main/java/com/hiqes/andele/RequestOwnerAppCompatActivity.java (hunk 3 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/RequestOwnerAppCompatActivity.java:24]: Both are now supported

File: build.gradle (hunk 1 above)
Comment (rspl-kashyap) [reply build.gradle]: Done

File: src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,4 +1,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.hiqes.andele" >
+    xmlns:tools="http://schemas.android.com/tools"
+    package="com.hiqes.andele">
+
+    <application
+        android:appComponentFactory="android.support.v4.app.CoreComponentFactory"
```
Comment (hiq-larryschiefer) [inline src/main/AndroidManifest.xml]: Why is this necessary?  What implication will this have for applications which use this library?

File: src/main/java/com/hiqes/andele/Andele.java [hunk 5]
```diff
@@ -208,8 +226,8 @@ public static void checkAndExecute(Activity activity, ProtectedAction action) {
      *                   the ActionCallback to execute when granted, etc.
      */
     @SuppressWarnings({"WeakerAccess", "unused"})
-    public static void checkAndExecute(AppCompatActivity activity, ProtectedAction action) {
-        checkAndExecute(new RequestOwnerAppCompatActivity(activity), action);
+    public static void checkAndExecute(android.support.v7.app.AppCompatActivity activity, ProtectedAction action) {
+        checkAndExecute(new RequestOwnerSupportAppCompatActivity(activity), action);
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/Andele.java]: It would be better to not change the type here since there could be codebases which already use it.  Introducing the new type for AndroidX above is ok, but keeping the existing type is needed for compatibility.

File: src/main/java/com/hiqes/andele/Andele.java [hunk 6]
```diff
@@ -590,7 +667,7 @@ public static void startSettingsApp(Fragment fragment) {
      * @param fragment   The Fragment in the running state making the request.
      */
     @SuppressWarnings({"WeakerAccess", "unused"})
-    public static void startSettingsApp(android.support.v4.app.Fragment fragment) {
+    public static void startSettingsApp(androidx.fragment.app.Fragment fragment) {
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/Andele.java:593]: Will this still work with older codebases using the `v4.app.Fragment`?

File: src/main/java/com/hiqes/andele/Andele.java [hunk 7]
```diff
@@ -765,6 +842,8 @@ public void handleMessage(Message msg) {
                     //  education info about, so do the request.
                     if (!skipAsk) {
                         doRequest(msg.arg1);
+                    } else {
+                        sReqMgr.removeRequest(msg.arg1);
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/Andele.java]: Add a comment describing why the request is being removed.

Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/Andele.java]: Done

File: src/main/java/com/hiqes/andele/Andele.java (hunk 6 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/Andele.java:593]: My mistake. Added the overloaded method.

File: src/main/java/com/hiqes/andele/RequestOwnerAndroidXAppCompatActivity.java [hunk 8]
```diff
@@ -96,11 +96,11 @@ public boolean shouldShowRequestPermissionRationale(String permission) {
 
     @Override
     public boolean isSameOwner(RequestOwner otherOwner) {
-        boolean                         ret = false;
-        RequestOwnerAppCompatActivity   otherOwnerAct = null;
+        boolean                               ret           = false;
+        RequestOwnerAndroidXAppCompatActivity otherOwnerAct = null;
 
         try {
-            otherOwnerAct = (RequestOwnerAppCompatActivity)otherOwner;
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/RequestOwnerAndroidXAppCompatActivity.java]: This likely needs to cascade between casts to `RequestOwnerAppCompatActivity` and `RequestOwnerAndroidXAppCompatActivity`.  Both cases need to be covered, not just AndroidX.

Comment (hiq-larryschiefer) [reply src/main/java/com/hiqes/andele/RequestOwnerAndroidXAppCompatActivity.java]: Actually, I now see that you refactored this class completely.  We will need 2 different classes: one for the old app compat and one for Android X app compat.

File: src/main/java/com/hiqes/andele/RequestOwnerSupportAppCompatActivity.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.support.v7.app.AppCompatActivity;
+import android.view.View;
+
+import java.lang.ref.WeakReference;
+
+class RequestOwnerSupportAppCompatActivity extends RequestOwner {
```
Comment (hiq-larryschiefer) [inline src/main/java/com/hiqes/andele/RequestOwnerSupportAppCompatActivity.java]: This should be the new class for Android X app compat, the existing one should have been left as-is.

File: src/main/java/com/hiqes/andele/Andele.java (hunk 5 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/Andele.java]: Changes done as discussed.

File: src/main/java/com/hiqes/andele/RequestOwnerAndroidXAppCompatActivity.java (hunk 8 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/RequestOwnerAndroidXAppCompatActivity.java]: Changes done as discussed.

File: src/main/java/com/hiqes/andele/RequestOwnerSupportAppCompatActivity.java (hunk 9 above)
Comment (rspl-kashyap) [reply src/main/java/com/hiqes/andele/RequestOwnerSupportAppCompatActivity.java]: Yes, Done.

File: build.gradle [hunk 10]
```diff
@@ -114,13 +118,18 @@ artifacts {
...
+    testImplementation 'org.mockito:mockito-core:2.19.0'
 
-    compile 'com.android.support:appcompat-v7:26.1.0'
-    compile 'com.android.support:design:26.1.0'
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'com.google.android.material:material:1.0.0'
 
-    androidTestCompile('com.android.support.test.espresso:espresso-core:2.2.2', {
+    implementation('com.android.support:appcompat-v7:28.0.0', {
+        exclude group: 'androidx.appcompat'
```
Comment (hiq-larryschiefer) [inline build.gradle]: My guess is that these `exclude` lines are what allows `com.android.support:appcompat-v7` to be used along side `androidx.appcompat:appcompat`, right?

Comment (rspl-kashyap) [reply build.gradle]: Yes. You suggested it on the Friday.

**Review by hiq-larryschiefer — APPROVED** [review]

**Review by hiq-larryschiefer — COMMENTED** [review]
You can actually just wait and rebase this branch against `master` once the other branch has been merged.

**Review by hiq-larryschiefer — APPROVED** [review]
